homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,261 @@
+                uid=task["id"],
+                rrule=str(build_rrule(task))[30:],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.DAILY and task["everyX"]
+            for recurrence in build_rrule(task).between(start_date, end_date, inc=True)
+            if (start := recurrence.date()) > today
+            or (start == today and not task["completed"])
+        ]
+
+
+class HabiticaRemindersCalendarEntity(HabiticaCalendarEntity):
+    """Habitica reminders calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.REMINDERS,
+        translation_key=HabiticaCalendar.REMINDERS,
+    )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        events_sorted = None
+        now = dt_util.now()
+        events = [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(hours=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=f"{task["id"]}_{reminder["id"]}",
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] in (HabiticaTaskType.DAILY, HabiticaTaskType.TODO)
+            and not task["completed"]
+            for reminder in task.get("reminders", [])
+            if (
+                start := datetime.fromisoformat(reminder["time"]).replace(
+                    tzinfo=dt_util.DEFAULT_TIME_ZONE
+                )
+            )
+            >= (now - timedelta(hours=1))
+        ]
+
+        events_sorted = sorted(
+            events,
+            key=lambda event: event.start,
+        )
+
+        # return the last reminder that went off in the last hour
+        if past_events := [
+            event
+            for event in events_sorted
+            if now >= event.start >= (now - timedelta(hours=1))
+        ]:
+            return past_events[-1]
+
+        # if there is no active reminder, move on to the next upcoming event
+        if upcoming_events := [event for event in events_sorted if event.start > now]:
+            return upcoming_events[0]
+        return None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        return [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(hours=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=f"{task["id"]}_{reminder["id"]}",
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] in (HabiticaTaskType.DAILY, HabiticaTaskType.TODO)
+            and not task["completed"]
+            for reminder in task.get("reminders", [])
+            if (
+                start_date
```

Comment:
There are two changes needed here i think to meet the spec described here  https://developers.home-assistant.io/docs/core/entity/calendar/#get-events

(1) the end date is exclusive.
(2) This need to check if the event end date is between requested start and end dates.

Reply:
The events don't have enddates, so it is not necessary to check if the enddate is in between.

Reply:
All calendar events have an end date. I understand the habitica tasks/events don't, but i'm talking about how its represented as a home assistant event.  They set an end date  ("start + timedelta(hours=1)".) This is not implementing the get events spec correctly "The start_date is the lower bound and applied to the event's end (exclusive)."

Reply:
I removed the reminders for now as there are still some other things i have to figure out about how they work in Habitica, the to-dos and dailies are all-day events so i think this doesn't apply to them.